---
title: "2_sc-seq_dasatinib_strong_qc"
author: "Kieran Redpath"
date: "2023-10-26"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Load packages
```{r, warning=FALSE}
# General
library(dplyr)
# For single cell analysis
library(Seurat)
library(patchwork)
library(celldex)
library(SingleR)
library(scater)
# For sc-type cell type annotation
library(HGNChelper)
library(openxlsx)
# sc-type functions from GitHub
source("https://raw.githubusercontent.com/IanevskiAleksandr/sc-type/master/R/gene_sets_prepare.R");
source("https://raw.githubusercontent.com/IanevskiAleksandr/sc-type/master/R/sctype_score_.R")
source("https://raw.githubusercontent.com/IanevskiAleksandr/sc-type/master/R/auto_detect_tissue_type.R")
# For plotting sc-type scores
lapply(c("ggraph","igraph","tidyverse", "data.tree"), library, character.only = T)
# For differential expression analysis
library(limma)
```

# Load data
```{r}
das.dat <- readRDS("~/R/sc-seq_mouse/data/scDrugMiceDas_annotated.RDS")
```

# Quality control
```{r, warning=FALSE}
# Examine proportion of mitochondrial features, as well as n features and counts per sample.
das.dat[["percent.mt"]] <- PercentageFeatureSet(das.dat, pattern = "^mt-")
VlnPlot(das.dat, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
# there are varying proportions of mt genes across samples

# FeatureScatter is typically used to visualize feature-feature relationships, but can be used
# for anything calculated by the object, i.e. columns in object metadata, PC scores etc.
plot.counts.vs.features <- FeatureScatter(das.dat, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
plot.counts.vs.features

# This line is adapted from the seurat tutorial with the goal of filtering out doublets (at least by filtering on counts), however our data uses probes to try and capture more features with the same amount of reads so may need to be treated differently. Should try filter out counts above something to remove outliers
# Original filtering (i.e. on poster)
# das.dat <- subset(das.dat, subset = nFeature_RNA > 200)

# Look for outliers to remove that are at least 3 MADS (mean absolute deviations) from the mean
high.percent.mt <- isOutlier(das.dat@meta.data[["percent.mt"]], nmads = 3, type = "higher")
table(high.percent.mt)
attr(high.percent.mt, "thresholds")[2]
# This would exclude xxx cells based on mtDNA percentage
low.high.features <- isOutlier(das.dat@meta.data[["nFeature_RNA"]], nmads = 3, type = "both")
table(low.high.features)
attr(low.high.features, "thresholds")
# This would exclude xxx cells based on nFeatures
low.high.counts <- isOutlier(das.dat@meta.data[["nCount_RNA"]], nmads = 3, type = "both")
table(low.high.counts)
attr(low.high.counts, "thresholds")
# This would exclude xxx cells based on nCounts

# On the following plot, many of the cells with very high proportions of mtDNA have very few features, indicating that they aren't good quality
plot.features.vs.percent.mt <- FeatureScatter(das.dat, feature1 = "nFeature_RNA", feature2 = "percent.mt")
plot.features.vs.percent.mt

# Try and plot some of these thresholds on the data to see what you're removing:
# Add high percent mt information to metadata
das.dat@meta.data[["high.percent.mt"]] <- high.percent.mt
# Plot percent mt against total counts
ggplot(das.dat@meta.data,  
       aes(x = nCount_RNA, y = percent.mt, color = high.percent.mt)) +
  geom_point() +
  labs(y = "Percentage mitochondrial transcripts", x = "Counts", title = "Mitochondrial transcripts")

# Plot percent mt across samples to see if any are especially high
ggplot(das.dat@meta.data,  
       aes(x = samp.ident.desc, y = percent.mt, color = samp.ident.desc)) +
  geom_jitter() +
  geom_violin() +
  geom_abline(intercept = attr(high.percent.mt, "thresholds")[2], slope = 0, color = "black") +
  geom_abline(intercept = 10, slope = 0, color = "black", lty = 2) +
  theme(legend.position = "none", axis.text.x = element_text(angle = 45, hjust = 1)) +
  labs(y = "% Mitochondrial counts", x = "Samples", title = "% Mitochondrial counts")

# Plot counts against samples to see which samples have more counts
ggplot(das.dat@meta.data,  
       aes(x = samp.ident.desc, y =nCount_RNA, color = samp.ident.desc)) +
  geom_jitter() +
  geom_violin() +
  geom_abline(intercept = attr(low.high.counts, "thresholds")[1], slope = 0, color = "black") +
  geom_abline(intercept = attr(low.high.counts, "thresholds")[2], slope = 0, color = "black") +
  theme(legend.position = "none", axis.text.x = element_text(angle = 45, hjust = 1)) +
  labs(y = "Counts", x = "Samples", title = "Counts")

# Plot features against samples to see which samples have more features
ggplot(das.dat@meta.data,  
       aes(x = samp.ident.desc, y =nFeature_RNA, color = samp.ident.desc)) +
  geom_jitter() +
  geom_violin() +
  geom_abline(intercept = attr(low.high.features, "thresholds")[1], slope = 0, color = "black") +
  geom_abline(intercept = attr(low.high.features, "thresholds")[2], slope = 0, color = "black") +
  geom_abline(intercept = 200, slope = 0, color = "black", lty = 2) +
  geom_abline(intercept = 7500, slope = 0, color = "black", lty = 2) +
  theme(legend.position = "none", axis.text.x = element_text(angle = 45, hjust = 1)) +
  labs(y = "Features", x = "Samples", title = "Features")

# Plot counts against percent mt, looking for a good cutoff
plot.counts.vs.percent.mt <- FeatureScatter(das.dat, feature1 = "nCount_RNA", feature2 = "percent.mt")#, fill.by = "high.percent.mt")
plot.counts.vs.percent.mt
```

# Filter data with thresholds based on the above plots
```{r}
# This line is adapted from the Seurat tutorial with the goal of filtering out doublets (at least by filtering on counts), however our data uses probes to try and capture more features with the same amount of reads so may need to be treated differently. This is very light filtering.
# das.dat <- subset(das.dat, subset = nFeature_RNA > 200)

# Tighter filtering, but still weaker than in the strong_qc document and output
das.dat <- subset(das.dat, subset = nFeature_RNA > 200 & nCount_RNA < 25000 & percent.mt < 5)
```

# Normalisation
```{r}
# Using default settings
das.dat <- NormalizeData(das.dat, normalization.method = "LogNormalize", scale.factor = 10000)
```

# Quick Side-bar - determine if das only cluster 8 still exists in the same way with the same markers between das treated and all control samples - not sure if you need to worry about this, but if you do then you need to also load mouse.dat in and normalise it etc
```{r}
# # Find cluster markers to see if the same cluster is still present in the full dataset including the other control samples
# # Subset data
# das.vs.control.dat <- subset(mouse.dat, subset = treat.ident == c("Das - drug", "Das - control", "Ent - control", "Mir - control"))
# 
# # Dimensions
# das.vs.control.dat <- JackStraw(das.vs.control.dat, num.replicate = 100)
# das.vs.control.dat <- ScoreJackStraw(das.vs.control.dat, dims = 1:20)
# 
# JackStrawPlot(das.vs.control.dat, dims = 1:20)
# ElbowPlot(das.vs.control.dat, ndims = 50)
# das.vs.control.dat <- FindNeighbors(das.vs.control.dat, dims = 1:16)
# das.vs.control.dat <- FindClusters(das.vs.control.dat, resolution = 0.6)
# das.vs.control.dat <- RunUMAP(das.vs.control.dat, dims = 1:16)
# DimPlot(das.vs.control.dat, reduction = "umap", label = T)
# DimPlot(das.vs.control.dat, reduction = "umap", split.by = "treat.ident", label = T)
# DimPlot(das.vs.control.dat, reduction = "umap", group.by = "samp.ident")
# DimPlot(das.vs.control.dat, reduction = "umap", split.by = "samp.ident", label = T)
# 
# # Muc13 and Dmbt1, named markers for OG cluster 8
# DimPlot(das.vs.control.dat, reduction = "umap", label = T, split.by = "samp.ident.desc")
# FeaturePlot(das.vs.control.dat, features = "Muc13", split.by = "samp.ident.desc")
# FeaturePlot(das.vs.control.dat, features = "Dmbt1", split.by = "samp.ident.desc")
# 
# # Dmbt1 isn't as strong of a marker for cluster 8, which is why it isn't on the list, but is expressed here
# # Alpi and Akp3 (human/mouse orthologs), top 2 markers for cluster 8
# FeaturePlot(das.vs.control.dat, features = "Akp3", split.by = "samp.ident")
# FeaturePlot(das.vs.control.dat, features = "Akp3", split.by = "samp.ident")
# # Ada and Adh6a, top third and fourth markers for cluster 8
# FeaturePlot(das.vs.control.dat, features = "Ada", split.by = "samp.ident")
# FeaturePlot(das.vs.control.dat, features = "Adh6a", split.by = "samp.ident")
# 
# # Find das.vs.control.dat markers for cluster 3, which seems similar to cluster 8:
# das_vs_control.markers.3 <- FindMarkers(das.vs.control.dat, ident.1 = 3) %>% .[order(.$avg_log2FC, decreasing = TRUE ),]
# write.csv(das_vs_control.markers.3, file = "2_output_strong_qc/das_vs_control_markers_3.csv")
# 
# png(file = "2_output_strong_qc/das_vs_all_controls.png", width = 4000, height = 1000, units = "px")
# DimPlot(das.vs.control.dat, reduction = "umap", label = T, split.by = "samp.ident.desc")
# dev.off()
# # These markers appear to be expressed in the same cluster in each group except the two das controls, so it looks like it could just be cells that weren't captured by chance in the das controls.
# # Conversely, it could be more a matter of: You're bound to capture some other cells that express certain genes the harder you look (i.e. more cells) and it could just be a matter of the larger size of those datasets.
```

# Feature selection
```{r}
# Using default settings to find the most variable features to inform our clustering
das.dat <- FindVariableFeatures(das.dat, selection.method = "vst", nfeatures = 2000)

# Identify the 10 most highly variable genes
top10 <- head(VariableFeatures(das.dat), 10)

# # Plot all genes by variance, labelling the top 10
variable.genes <- VariableFeaturePlot(das.dat)
plot.variable.genes.lab <- LabelPoints(plot = variable.genes, points = top10, repel = TRUE)
plot.variable.genes.lab
```

# Dimensional reduction
```{r}
# Apply linear transformation prior to PCA and other downstream analyses
all.genes <- rownames(das.dat)
das.dat <- ScaleData(das.dat, features = all.genes)

# Perform linear dimensional reduction
das.dat <- RunPCA(das.dat, features = VariableFeatures(object = das.dat))

# Examine and visualize PCA results a few different ways
# Provide lists of positively and negatively regulated genes per PC
print(das.dat[["pca"]], dims = 1:5, nfeatures = 5)
# Shows clearly up and down regulated genes in PCs
VizDimLoadings(das.dat, dims = 1:2, reduction = "pca")
# Plots first two PCs and shows all genes in each sample
DimPlot(das.dat, reduction = "pca")
# Can be used to inform which PCs to use for downstream analysis (but to do so you should plot all PCs)
DimHeatmap(das.dat, dims = 1, cells = 500, balanced = TRUE)
```

# Determine dimensionality of the data
```{r}
# This can help inform the number of PCs to use by plotting the distribution of p-values for each PC with a uniform distribution
# This bit can take awhile
das.dat <- JackStraw(das.dat, num.replicate = 100)
das.dat <- ScoreJackStraw(das.dat, dims = 1:20)

# The JackStrawPlot() function provides a visualization tool for comparing the distribution of p-values for each PC with a uniform distribution (dashed line). ‘Significant’ PCs will show a strong enrichment of features with low p-values (solid curve above the dashed line).
JackStrawPlot(das.dat, dims = 1:20)
# Alternatively, PCs can be ranked in a less computationally-intensive way using the ElbowPlot function
ElbowPlot(das.dat, ndims = 50)
# Looking at the ElbowPlot, since the JackStrawPlot is confusing, looks like about 11 PCs is pretty good
# More notes on these methods can be found in the Seurat guide
```

# Cluster the cells and visualise dimensions with UMAP or tSNE
```{r}
# Construct a K nearest neighbours graph based on the distance between points in PCA space and refine edge weights between cells with Jaccard similarity. Using the first 20 PCs here.
das.dat <- FindNeighbors(das.dat, dims = 1:20)

# Cluster the cells using Louvain algorithm. Setting resolution between 0.4 and 1.2 for a 3K cells dataset is recommended in the documentation, but optimal resolution often increases for larger datasets.
das.dat <- FindClusters(das.dat, resolution = 0.6)

# Look at cluster IDs of the first 5 cells
head(Idents(das.dat), 5)

# Run and plot UMAP for dimensional reduction, first for all cells, second by treatment and third by sample names to show that the cells are clustered more by type than sample, and finally split by sample names to show the distribution of different clusters across each sample
das.dat <- RunUMAP(das.dat, dims = 1:20)
DimPlot(das.dat, reduction = "umap", label = T)
DimPlot(das.dat, reduction = "umap", split.by = "treat.ident", label = T)
DimPlot(das.dat, reduction = "umap", group.by = "samp.ident.plain")
DimPlot(das.dat, reduction = "umap", split.by = "samp.ident.plain", label = T)
```

# Attempt to assign some cell types automatically using sc-type and its database

# Create database of reference data and gene sets for sc-type
```{r}
# Load database
database <- "https://raw.githubusercontent.com/IanevskiAleksandr/sc-type/master/ScTypeDB_full.xlsx";
# Define tissue type
tissue <- "Stomach"
# Guess tissue type
# tissue_guess = auto_detect_tissue_type(path_to_db_file = database, seuratObject = das.dat, scaled = FALSE, assay = "RNA")
# if scaled = TRUE, make sure the data is scaled, as seuratObject[[assay]]@scale.data is used. If you just created a Seurat object, without any scaling and normalization, set scaled = FALSE, seuratObject[[assay]]@counts will be used         
# Prepare gene sets
gs_list <- gene_sets_prepare(database, tissue)
```

# Find Marker Genes for each cluster with Seurat and sctype
```{r}
# Find markers with Seurat
das.markers <- FindAllMarkers(das.dat, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25) %>% .[order(.$avg_log2FC, decreasing = TRUE ),]

# Assign cell types to each cluster using sctype
es.max = sctype_score(scRNAseqData = das.dat[["RNA"]]@scale.data, scaled = TRUE, 
                      gs = gs_list$gs_positive, gs2 = gs_list$gs_negative) 

# Merge by cluster
cl_results = do.call("rbind", lapply(unique(das.dat@meta.data$seurat_clusters), function(cl){
    es.max.cl = sort(rowSums(es.max[ ,rownames(das.dat@meta.data[das.dat@meta.data$seurat_clusters==cl, ])]), decreasing = !0)
    head(data.frame(cluster = cl, type = names(es.max.cl), scores = es.max.cl, ncells = sum(das.dat@meta.data$seurat_clusters==cl)), 10)
}))
sctype_scores = cl_results %>% group_by(cluster) %>% top_n(n = 1, wt = scores)  

# set low-confident (low ScType score) clusters to "unknown"
sctype_scores$type[as.numeric(as.character(sctype_scores$scores)) < sctype_scores$ncells/4] = "Unknown"
print(sctype_scores[,1:3])
#sctype_score uses positive (gs) and negative (gs2) markers. If you don't want to use negative markers, set gs2 to NULL

# Assign types and plot UMAP
das.dat@meta.data$type.sctype = ""
for(j in unique(sctype_scores$cluster)){
  cl_type = sctype_scores[sctype_scores$cluster==j,]; 
  das.dat@meta.data$type.sctype[das.dat@meta.data$seurat_clusters == j] = as.character(cl_type$type[1])
}

# Bubble plot
# prepare edges
cl_results=cl_results[order(cl_results$cluster),]; edges = cl_results; edges$type = paste0(edges$type,"_",edges$cluster); edges$cluster = paste0("cluster ", edges$cluster); edges = edges[,c("cluster", "type")]; colnames(edges) = c("from", "to"); rownames(edges) <- NULL

# prepare nodes
nodes_lvl1 = sctype_scores[,c("cluster", "ncells")]; nodes_lvl1$cluster = paste0("cluster ", nodes_lvl1$cluster); nodes_lvl1$Colour = "#f1f1ef"; nodes_lvl1$ord = 1; nodes_lvl1$realname = nodes_lvl1$cluster; nodes_lvl1 = as.data.frame(nodes_lvl1);
nodes_lvl2 = c(); 
ccolss= c("#5f75ae","#92bbb8","#64a841","#e5486e","#de8e06","#eccf5a","#b5aa0f","#e4b680","#7ba39d","#b15928","#ffff99", "#6a3d9a","#cab2d6","#ff7f00","#fdbf6f","#e31a1c","#fb9a99","#33a02c","#b2df8a","#1f78b4","#a6cee3", "#03dbfc")
for (i in 1:length(unique(cl_results$cluster))){
  dt_tmp = cl_results[cl_results$cluster == unique(cl_results$cluster)[i], ]; nodes_lvl2 = rbind(nodes_lvl2, data.frame(cluster = paste0(dt_tmp$type,"_",dt_tmp$cluster), ncells = dt_tmp$scores, Colour = ccolss[i], ord = 2, realname = dt_tmp$type))
}

nodes = rbind(nodes_lvl1, nodes_lvl2); nodes$ncells[nodes$ncells<1] = 1;
files_db = openxlsx::read.xlsx(database)[,c("cellName","shortName")]; files_db = unique(files_db); nodes = merge(nodes, files_db, all.x = T, all.y = F, by.x = "realname", by.y = "cellName", sort = F)
nodes$shortName[is.na(nodes$shortName)] = nodes$realname[is.na(nodes$shortName)]; nodes = nodes[,c("cluster", "ncells", "Colour", "ord", "shortName", "realname")]

# Remove inexplicably duplicated rows, where shortname is mesothelial, not mesothelial_cells
nodes <- nodes[nodes$shortName != "Mesothelial", ]

mygraph <- graph_from_data_frame(edges, vertices=nodes)
# Doesn't work due to duplicate vertex names (first column of nodes). This is because for some reason each of the mesothelial cell clusters is duplicated, with one including the short name "mesothelial", and the other "mesothelial_cells"

# Make the graph
sctype.per.cluster <- ggraph(mygraph, layout = 'circlepack', weight=I(ncells)) + 
  geom_node_circle(aes(filter=ord==1,fill=I("#F5F5F5"), colour=I("#D3D3D3")), alpha=0.9) + geom_node_circle(aes(filter=ord==2,fill=I(Colour), colour=I("#D3D3D3")), alpha=0.9) +
  theme_void() + geom_node_text(aes(filter=ord==2, label=realname, colour=I("#ffffff"), fill="white", repel = !1, parse = T, size = I(log(ncells,25)*1.2)))+ geom_node_label(aes(filter=ord==1,  label=realname, colour=I("#000000"), size = I(3), fill="white", parse = T), repel = !0, segment.linetype="dotted")
 
# Use gridExtra to print these graphs next to each other, with the UMAP annotated with original cluster ID's, not annotated cell types
png(filename = "2_output_strong_qc/type.sctype_and_types_per_cluster.png", width = 2000, height = 1000, units = "px")
gridExtra::grid.arrange(DimPlot(das.dat, reduction = "umap", label = TRUE, repel = TRUE, cols = ccolss, group.by = "type.sctype"), sctype.per.cluster, ncol = 2)
dev.off()

# UMAP with sctype types and treatment identification
png(filename = "2_output_strong_qc/type.sctype_by_treat.ident.png", width = 1500, height = 1000, units = "px")
DimPlot(das.dat, reduction = "umap", split.by = "treat.ident", group.by = "type.sctype", label = T)
dev.off()
DimPlot(das.dat, reduction = "umap", split.by = "treat.ident", group.by = "type.sctype", label = T)
# UMAP with seurat cluster idents
DimPlot(das.dat, reduction = "umap", label = TRUE, repel = TRUE)
# UMAP with seurat cluster idents and treatment identification
DimPlot(das.dat, reduction = "umap", label = TRUE, repel = TRUE, split.by = "treat.ident")
png(filename = "2_output_strong_qc/seurat_cluster_by_treat.ident.png", width = 1500, height = 1000, units = "px")
DimPlot(das.dat, reduction = "umap", label = TRUE, repel = TRUE, split.by = "treat.ident")
dev.off()
# sctype annotations per cluster
sctype.per.cluster
```

# In this data, it appears that cluster 8 is only present in drugged cells. Let's explore markers for cluster 8
```{r}
# Print the top 10 markers for cluster 8 ("MUC13_DMBT1 positive cells) vs all other cells
das.markers.8 <- das.markers %>%
  group_by(cluster) %>%
  filter(cluster == 8)

# These probably aren't super useful.
# # Save the top 100 positive marker genes for cluster 8 specifically 
# das.markers.8.pos <- das.markers %>%
#   group_by(cluster) %>%
#   slice_max(n = 100, order_by = avg_log2FC) %>% 
#   filter(cluster == 8) %>% .$gene
# write(das.markers.8.pos, "2_output_strong_qc/das.markers.8.pos.txt")

# Save the top 100 negative marker genes for cluster 8 specifically (DON'T CURRENTLY HAVE ANY NEG, ALL THIS CODE DOES IS SAVE THE BOTTOM 100 POS)
# das.markers.8.neg <- das.markers %>%
#   group_by(cluster) %>%
#   slice_min(n = 100, order_by = avg_log2FC) %>% 
#   filter(cluster == 8) %>% .$gene
# write(das.markers.8.neg, "2_output_strong_qc/das.markers.8.neg.txt")

# Cluster 8 cells are annotated as Muc13, Dmbt1 positive, so let's look at the expression of these genes in each sample and see where they lie
# Add a meta.data value to Seurat object for whether or not cells are part of the MUC13_DMBT1 positive cells cluster
isMUC13_DMBT1pos <- FetchData(das.dat, vars = "type.sctype")
isMUC13_DMBT1pos$TF <- isMUC13_DMBT1pos$type.sctype == 'MUC13_DMBT1 positive cells'
das.dat <- AddMetaData(das.dat, metadata = isMUC13_DMBT1pos$TF, col.name = "isMUC13_DMBT1pos")

# Muc13 and Dmbt1, named markers for cluster 8
FeaturePlot(das.dat, features = c("Muc13", "Dmbt1"), split.by = "treat.ident", shape.by = "isMUC13_DMBT1pos")
# Dmbt1 isn't as strong of a marker for cluster 8, which is why it isn't on the list, but is expressed here
# Alpi and Akp3 (human/mouse orthologs), top 2 markers for cluster 8
FeaturePlot(das.dat, features = c("Alpi", "Akp3"), split.by = "treat.ident", shape.by = "isMUC13_DMBT1pos")
# Ada and Adh6a, top third and fourth markers for cluster 8
FeaturePlot(das.dat, features = c("Ada", "Adh6a"), split.by = "treat.ident", shape.by = "isMUC13_DMBT1pos")

# Let's also look at Cdh1 and Trp53 - If we're looking at cancer cells (but only the responding ones) in cluster 8, we'd hope these genes weren't expressed strongly there
# Cdh1
FeaturePlot(das.dat, features = "Cdh1", split.by = "samp.ident.plain", shape.by = "isMUC13_DMBT1pos")
#Trp53
FeaturePlot(das.dat, features = "Trp53", split.by = "samp.ident.plain", shape.by = "isMUC13_DMBT1pos")
# Cd44 (since Cdh1/Trp53 KO cells should be expressing Cd44
FeaturePlot(das.dat, features = "Cd44", split.by = "samp.ident.plain", shape.by = "isMUC13_DMBT1pos")
# Krt14, which is underexpressed with Cdh1 loss according to a paper from the lab
FeaturePlot(das.dat, features = "Krt14", split.by = "samp.ident.plain", shape.by = "isMUC13_DMBT1pos")
# -------------------------------------------------------------------------------------------------------------
# Do these all on Vlnplots, which mean you don't need to shape.by cluster and look way better
# Muc13 and Dmbt1, named markers for cluster 8
VlnPlot(das.dat, features = c("Muc13", "Dmbt1"),  group.by = "seurat_clusters")
# Dmbt1 isn't as strong of a marker for cluster 8, which is why it isn't on the list, but is expressed here
# Alpi and Akp3 (human/mouse orthologs), top 2 markers for cluster 8
VlnPlot(das.dat, features = c("Alpi", "Akp3"),  group.by = "seurat_clusters")
# Ada and Adh6a, top third and fourth markers for cluster 8
VlnPlot(das.dat, features = c("Ada", "Adh6a"),  group.by = "seurat_clusters")

# Let's also look at Cdh1 and Trp53 - If we're looking at cancer cells (but only the responding ones) in cluster 8, we'd hope these genes weren't expressed strongly there
# Cdh1
VlnPlot(das.dat, features = "Cdh1",  group.by = "seurat_clusters")
#Trp53
VlnPlot(das.dat, features = "Trp53",  group.by = "seurat_clusters")
# Epcam (epithelial cell marker)
VlnPlot(das.dat, features = "Epcam",  group.by = "seurat_clusters")
# Cd44 (since Cdh1/Trp53 KO cells should be expressing Cd44
VlnPlot(das.dat, features = "Cd44",  group.by = "seurat_clusters")
# Krt14, which is underexpressed with Cdh1 loss according to a paper from the lab
VlnPlot(das.dat, features = "Krt14",  group.by = "seurat_clusters") 
# Plot Olfm4, Cdh17, Krt20, Muc5ac, Lgals4, Akr1B10, and Reg4, all of which have been found decently expressed in human gastric cancer
VlnPlot(das.dat, features = c("Olfm4", "Cdh17", "Krt20", "Muc5ac"), group.by = "seurat_clusters")

VlnPlot(das.dat, features = c("Lgals4", "Akr1B10", "Reg4"), group.by = "type.sctype")

# ---------------------------------------------------------------------------------------------------------------

# Try to find other clusters that might represent cancer cells by plotting Cdh1 and Trp53 across all samples (Dmbt1 and Muc13 are just for fun, they probably aren't important in cancer as a whole), as this should be consistent
# VlnPlot
VlnPlot(das.dat, features = c("Cdh1", "Trp53", "Cd44", "Dmbt1", "Muc13"),  group.by = "seurat_clusters")
# Also plot Gkn2, Tff1, Tff2,normal gastric mucosal epithelial markers https://www.cell.com/cell-reports/pdfExtended/S2211-1247(22)01332-8
VlnPlot(das.dat, features = c("Cdh1", "Gkn2", "Tff1", "Tff2"), group.by = "seurat_clusters")
# Plot some good enterocyte markers (the top 5 on Panglao, plus Apobec1 for fun) to see if they're more highly expressed in cluster 8
VlnPlot(das.dat, features = c("Slc10a2", "Alpi", "Krt20", "Si", "Rnf186", "Klf5", "Apobec1"),  group.by = "seurat_clusters")
# Plot Msmb, a gene identified as a marker for signet ring cells in a paper file:///C:/Users/axonn/OneDrive/University/PhD/Papers%20etc/Single-cell%20analysis%20of%20gastric%20signet%20ring%20cell%20carcinoma.pdf
VlnPlot(das.dat, features = "Msmb", group.by = "type.sctype")
FeaturePlot(das.dat, features = "Msmb")

#FeaturePlot
FeaturePlot(das.dat, features = c("Cdh1", "Trp53", "Cd44", "Dmbt1", "Muc13"))
# Also plot Gkn2, Tff1, Tff2,normal gastric mucosal epithelial markers https://www.cell.com/cell-reports/pdfExtended/S2211-1247(22)01332-8
FeaturePlot(das.dat, features = c("Cdh1", "Gkn2", "Tff1", "Tff2"))
# Also plot  some markers of apoptosis to see if cluster 8 is apoptosing
# CD95 (Fas), CD40, CD49d (Itga4) and CD11a, plus some others
FeaturePlot(das.dat, features = c("Fas", "Cd40", "Itga4", "Itgal", "Fasl", "Bcl2", "Nfkb1"))
# None of these cluster obviously to cluster 8 (or anywhere really)

# Featureplot proliferation markers
FeaturePlot(das.dat, features = c("Trp53", "Cdh1", "Mki67"))


DimPlot(das.dat, reduction = "umap", label = TRUE, repel = TRUE)
DimPlot(das.dat, reduction = "umap", label = TRUE, repel = TRUE, group.by = 'type.sctype')
# Cells clustered as and "Myeloid cells" (cluster 1), "Lymphatic endothelial cells" (cluster 2), and "Mesothelial cells" (cluster 4) are also lacking in both Cdh1 and Trp53 (mainly Cdh1), So may also represent cancer (or just cells that don't normally express these genes, such as non-epithelial cells)
# Checking the annotation of other cell type clusters too, to see if they might be cancerous (especially when multiple clusters have been labelled with the same cell types)
# What are the markers of cluster 0 ("Goblet cells")? 282 cells
das.markers.0 <- das.markers %>%
  group_by(cluster) %>%
  filter(cluster == 0)

# What are the markers of cluster 1 ("Myeloid cells")? 211 cells - Cdh1 and Tp53 negative, mostly
das.markers.1 <- das.markers %>%
  group_by(cluster) %>%
  filter(cluster == 1)

# What are the markers of cluster 2 ("Lymphatic endothelial cells")? 205 cells - Cdh1 and Tp53 negative, mostly
das.markers.2 <- das.markers %>%
  group_by(cluster) %>%
  filter(cluster == 2)

# What are the markers of cluster 3 ("Goblet cells")? 186 cells
das.markers.3 <- das.markers %>%
  group_by(cluster) %>%
  filter(cluster == 3)
 
# What are the markers of cluster 4 ("Mesothelial cells")? 155 cells - Cdh1 and Tp53 negative, mostly
das.markers.4 <- das.markers %>%
  group_by(cluster) %>%
  filter(cluster == 4)

# What are the markers of cluster 5 ("Lymphoid cells")? 143 cells
das.markers.5 <- das.markers %>%
  group_by(cluster) %>%
  filter(cluster == 5)

# What are the markers of cluster 6 ("Stromal cells")? 116 cells
das.markers.6 <- das.markers %>%
  group_by(cluster) %>%
  filter(cluster == 6)

# What are the markers of cluster 7 ("ENS neurons")? 106 cells
das.markers.7 <- das.markers %>%
  group_by(cluster) %>%
  filter(cluster == 7)

# What are the markers of cluster 9 ("Stromal cells")? 51 cells
das.markers.9 <- das.markers %>%
  group_by(cluster) %>%
  filter(cluster == 9)

# What are the markers of cluster 10 ("Parietal and chief cells")? 49 cells
das.markers.10 <- das.markers %>%
  group_by(cluster) %>%
  filter(cluster == 10)
```

# After a very thorough literature search of the top 5 positive marker genes (by log2 FC) for each cluster, re-label each cluster with your best guesses for identity
```{r}
type.manual.v1 <- c("Foveolar cells", "Macrophages", "Undetermined (lymphatic endothelial cells)", "Pyloric gland/neck cells", "Undetermined (serosal cells)", "Lymphoid cells", "Undetermined (fibroblasts)", "Pancreatic cells", "Enterocytes", "Stromal cells", "Parietal cells")
names(type.manual.v1) <- levels(das.dat)
das.dat <- RenameIdents(das.dat, type.manual.v1)
das.dat@meta.data[["type.manual.v1"]] <- das.dat@active.ident

png("2_output_strong_qc/annotedclusterids.png", height = 1000, width = 2000, units = "px")
DimPlot(das.dat, reduction = "umap", label = TRUE, repel = TRUE, label.size = 7, pt.size = 4, group.by = "type.manual.v1", split.by = "treat.ident") + NoLegend() + theme(plot.title = element_text(size = 20), plot.caption = element_text(size = 50))
dev.off()

png("2_output_strong_qc/VlnPlotCdh1.png", height = 1000, width = 2000, units = "px", res = 0.4, pointsize = 300)
VlnPlot(das.dat, features = "Cdh1", pt.size = 1) + NoLegend() +theme(axis.text.x=element_text(size=30))
dev.off()

VlnPlot(das.dat, features = "Cdh1") + NoLegend()

jpeg("2_output_strong_qc/VlnPlotCdh1.jpeg", height = 1000, width = 2000, units = "px")
VlnPlot(das.dat, features = "Cdh1") + NoLegend()
dev.off()

png("2_output_strong_qc/annotedclusteridsall.png", height = 1000, width = 2000, units = "px", pointsize = 20)
DimPlot(das.dat, reduction = "umap", label = TRUE, pt.size = 0.8, group.by = "type.manual.v1", split.by = "samp.ident.desc") + NoLegend()
dev.off()

png("2_output_strong_qc/type.sctype.png", height = 1000, width = 2000, units = "px", pointsize = 20)
DimPlot(das.dat, reduction = "umap", label = TRUE, pt.size = 0.8, group.by = "type.sctype", split.by = "treat.ident") + NoLegend()
dev.off()

# Plot some good enterocyte markers (the top 5 on Panglao, plus Apobec1 for fun) to see if they're more highly expressed in cluster 8
VlnPlot(das.dat, features = c("Slc10a2", "Alpi", "Krt20", "Si", "Rnf186", "Klf5", "Apobec1"))
# Could be intestinal enterocytes? Still kind of unsure

```

```{r}
# Cdh1 and what we think das is targeting (and also epithelial Ddr1 vs mesenchymal Ddr2)
FeaturePlot(das.dat, features = c("Cdh1", "Src"), split.by = "treat.ident")
FeaturePlot(das.dat, features = c("Ddr1", "Ddr2"), split.by = "treat.ident")
# AKt3 esp. was significantly correlated with CDH1
FeaturePlot(das.dat, features = c("Akt1", "Akt2", "Akt3"), split.by = "treat.ident")
FeaturePlot(das.dat, features = c("Pik3r1", "Mapk1"), split.by = "treat.ident")

# Genes correlated with AKT3 in TCGA and GEO that have been previously implicated in GC (also includes Ddr2)
FeaturePlot(das.dat, features = c("Cdh11", "Msrb3","Fstl1"), split.by = "treat.ident")

# EGFR and HER2
FeaturePlot(das.dat, features = c("Egfr", "Erbb2"), split.by = "treat.ident")
# MAPK/ERK signalling pathway
FeaturePlot(das.dat, features = c("Creb1", "Atf1"), split.by = "treat.ident")
# Mtorc1 and Mtorc2 (in some signalling pathway or another, also Tk's)
FeaturePlot(das.dat, features = c("Mtor", "Rptor"), split.by = "treat.ident")
# Regulated by Foxo, which is phosphorylated by PI3K/Akt signalling
FeaturePlot(das.dat, features = c("Ar", "Erg", "Runx2"), split.by = "treat.ident")
# Erk signalling
FeaturePlot(das.dat, features = c("Ccnd1", "Puma"), split.by = "treat.ident")
FeaturePlot(das.dat, features = c("Bcl2l11", "Bcl2"), split.by = "treat.ident")
FeaturePlot(das.dat, features = c("Mcl1", "Bcl2l1"), split.by = "treat.ident")

# Chat GPT suggested genes for Pi3k/Akt signalling
FeaturePlot(das.dat, features = "Foxo1", split.by = "treat.ident")
FeaturePlot(das.dat, features = "Mtor", split.by = "treat.ident")
FeaturePlot(das.dat, features = "Ccnd1", split.by = "treat.ident")
FeaturePlot(das.dat, features = "Bcl2", split.by = "treat.ident")
FeaturePlot(das.dat, features = "Nfkb1", split.by = "treat.ident")
FeaturePlot(das.dat, features = "Gsk3b", split.by = "treat.ident")
FeaturePlot(das.dat, features = "Hif1a", split.by = "treat.ident")
FeaturePlot(das.dat, features = "Slc2a4", split.by = "treat.ident")
FeaturePlot(das.dat, features = "Myc", split.by = "treat.ident")
FeaturePlot(das.dat, features = "Trp53", split.by = "treat.ident")
FeaturePlot(das.dat, features = "Akt1", split.by = "treat.ident")
FeaturePlot(das.dat, features = "Akt2", split.by = "treat.ident")
FeaturePlot(das.dat, features = "Akt3", split.by = "treat.ident")

# Chat GPT suggested names for Ddr2 signalling
FeaturePlot(das.dat, features = "Mmp2", split.by = "treat.ident")
FeaturePlot(das.dat, features = "Mmp3", split.by = "treat.ident")
FeaturePlot(das.dat, features = "Col1a1", split.by = "treat.ident")
FeaturePlot(das.dat, features = "Tgfb1", split.by = "treat.ident")
FeaturePlot(das.dat, features = "Fos", split.by = "treat.ident")
FeaturePlot(das.dat, features = "Jun", split.by = "treat.ident")
FeaturePlot(das.dat, features = "Ccn2", split.by = "treat.ident")
FeaturePlot(das.dat, features = "Pdgfa", split.by = "treat.ident")
FeaturePlot(das.dat, features = "Il6", split.by = "treat.ident")
FeaturePlot(das.dat, features = "Igf1", split.by = "treat.ident")

DimPlot(das.dat, reduction = "umap", label = TRUE, repel = TRUE, group.by = 'type.manual.v1', split.by = "samp.ident")
```

# Plot some genes
```{r}
# Plotting markers of MUC13_DMBT1 positive cells that the reference database uses to classify cell types, except Slc5a12 which no-one expresses, and GuCa2b, Mir194-2, and Btnl3, which weren't found.
FeaturePlot(das.dat, features = c("Isx","Enpp7","Tm4sf20","Si","Slc17a4","Aldob","Cdx1","Dmbt1","Chp2","Krt20","Adra2a"))

```

# Cell type vs treatment
```{r}
# For all cell types identified in controls, what portion of all data is represented

# The same in treat

# Compare
```

# Look for differentially expressed genes between cluster 8 and the rest of the dataset
* This seems like the next most sensible thing to do, as cluster 8 is the only one present in both treated samples but not controls (although it/a similar cluster is present in controls for other drugs), so is potentially likely to be involved in drug response.
* If you have time, just do this controls vs treated, using the 10X differential expression with pseudobulking guide.

# To try?
* Generating all das markers including negative, and seeing if sc-type can annotate better with these. Likely won't make any difference as database doesn't have any negative markers for the stomach
* Could use violin plots for gene expression between clusters instead of dotplots since it's a bit more scientific and vibes better with the clusters/cell types and showing where they are (plus you don't have to worry about changing the shape of dots to define only a single cluster)

################################################################## Insanity - stick together groups to make a table
########### Double check the info on if scale data is the normalised data. Make groups for all of your Cdh1 negative cell types and run DE between the treatment and control

# Split up dasatanib data by treatment type for pseudo-bulking and DE analysis
```{r}
# Split das dat into two elements of a list
das.dat.list <- SplitObject(das.dat, split.by = "treat.ident")
# Don't re-normalise, or add any additional information, just go from here

# Create a table containing cell types in each dataset and run a chisq test to look for a difference between proportions of cell types in each, excluding Enterocytes
control.cell.counts <- das.dat.list[["Das - control"]]@meta.data[["type.manual.v1"]]
table(control.cell.counts)
dasatinib.cell.counts <- das.dat.list[["Das - drug"]]@meta.data[["type.manual.v1"]] %>% .[-6]
table(dasatinib.cell.counts)

# Run a Chi-squared test on the data and show a proportion table of cell types in each group.
cbind(table(control.cell.counts), table(dasatinib.cell.counts)) %>% chisq.test()
cbind(table(control.cell.counts), table(dasatinib.cell.counts)) %>% prop.table(.,2)


# Make objects of log normalised RNA expression
das.control_dat <- das.dat.list[["Das - control"]]@assays[["RNA"]]@data
das.dasatinib_dat <- das.dat.list[["Das - drug"]]@assays[["RNA"]]@data

# Check number of genes and cells in each dataset
das.dat.list[["Das - control"]]@assays[["RNA"]]@data %>% dim()
# 921 control cells
das.dat.list[["Das - drug"]]@assays[["RNA"]]@data %>% dim()
# 636 drugged cells

# Make objects of treatments per cell
das.control.treat <- das.dat.list[["Das - control"]]@meta.data[["treat.ident"]]
das.dasatinib.treat <- das.dat.list[["Das - drug"]]@meta.data[["treat.ident"]]

# Make objects of annotated cluster ids per cell
das.control.clusters <- das.dat.list[["Das - control"]]@meta.data[["type.manual.v1"]]
das.dasatinib.clusters <- das.dat.list[["Das - drug"]]@meta.data[["type.manual.v1"]]

# Make objects of cluster number per cell
das.control.numerical.clusters <- das.dat.list[["Das - control"]]@meta.data[["seurat_clusters"]]
das.dasatinib.numerical.clusters <- das.dat.list[["Das - drug"]]@meta.data[["seurat_clusters"]]

# Genes with positive logFC mean they're more highly expressed in the dasatinib group
```

# DE on "Macrophages"
```{r}
# Define the cell type you want to extract
cluster.type <- "Macrophages"
# Subset data and run differential expression analysis on macrophages
das.control.macrophage <- das.control_dat[, das.control.clusters == cluster.type]
das.dasatinib.macrophage <- das.dasatinib_dat[, das.dasatinib.clusters == cluster.type]
das.combined.macrophage <- cbind(das.control.macrophage, das.dasatinib.macrophage)
das.groups.tmp <- c(das.control.treat[das.control.clusters == cluster.type], das.dasatinib.treat[das.dasatinib.clusters == cluster.type])

# Define grouping variable more clearly for limma
group <- ifelse(das.groups.tmp == "Das - control", "control", "dasatinib")
# Create design matrix: looks good
design <- model.matrix(~group);
colnames(design) <- c("Mean", "ControlVsDasatinib")
design %>% head()
# Fit linear model
fit <- lmFit(as.matrix(das.combined.macrophage), design)
fit <- eBayes(fit)
tt.macrophage <- topTable(fit, coef = "ControlVsDasatinib", adjust = "BH", n = nrow(das.combined.macrophage))
# Sum number of significantly DE genes in tt
sum(tt.macrophage$adj.P.Val<0.05)
# Report names of significantly DE genes
rownames(tt.macrophage[tt.macrophage$adj.P.Val < 0.05,])
DE.macrophage <- rownames(tt.macrophage[tt.macrophage$adj.P.Val < 0.05,])
```

# DE on "Lymphatic endothelial cells"
```{r}
# Define the cell type you want to extract
cluster.type <- "Undetermined (lymphatic endothelial cells)"
# Subset data and run differential expression analysis on lymphatic endothelial cells
das.control.lymphatic <- das.control_dat[, das.control.clusters == cluster.type]
das.dasatinib.lymphatic <- das.dasatinib_dat[, das.dasatinib.clusters == cluster.type]
das.combined.lymphatic <- cbind(das.control.lymphatic, das.dasatinib.lymphatic)
das.groups.tmp <- c(das.control.treat[das.control.clusters == cluster.type], das.dasatinib.treat[das.dasatinib.clusters == cluster.type])

# Define grouping variable more clearly for limma
group <- ifelse(das.groups.tmp == "Das - control", "control", "dasatinib")
# Create design matrix: looks good
design <- model.matrix(~group);
colnames(design) <- c("Mean", "ControlVsDasatinib")
design %>% head()
# Fit linear model
fit <- lmFit(as.matrix(das.combined.lymphatic), design)
fit <- eBayes(fit)
tt.lymphatic <- topTable(fit, coef = "ControlVsDasatinib", adjust = "BH", n = nrow(das.combined.lymphatic))
# Sum number of significantly DE genes in tt
sum(tt.lymphatic$adj.P.Val<0.05)
# Report names of significantly DE genes
rownames(tt.lymphatic[tt.lymphatic$adj.P.Val < 0.05,])
DE.lymphatic <- rownames(tt.lymphatic[tt.lymphatic$adj.P.Val < 0.05,])
```

# DE on "Serosal cells"
```{r}
# Define the cell type you want to extract
cluster.type <- "Undetermined (serosal cells)"
# Subset data and run differential expression analysis on serosal cells
das.control.serosal <- das.control_dat[, das.control.clusters == cluster.type]
das.dasatinib.serosal <- das.dasatinib_dat[, das.dasatinib.clusters == cluster.type]
das.combined.serosal <- cbind(das.control.serosal, das.dasatinib.serosal)
das.groups.tmp <- c(das.control.treat[das.control.clusters == cluster.type], das.dasatinib.treat[das.dasatinib.clusters == cluster.type])

# Define grouping variable more clearly for limma
group <- ifelse(das.groups.tmp == "Das - control", "control", "dasatinib")
# Create design matrix: looks good
design <- model.matrix(~group);
colnames(design) <- c("Mean", "ControlVsDasatinib")
design %>% head()
# Fit linear model
fit <- lmFit(as.matrix(das.combined.serosal), design)
fit <- eBayes(fit)
tt.serosal <- topTable(fit, coef = "ControlVsDasatinib", adjust = "BH", n = nrow(das.combined.serosal))
# Sum number of significantly DE genes in tt
sum(tt.serosal$adj.P.Val<0.05)
# Report names of significantly DE genes
rownames(tt.serosal[tt.serosal$adj.P.Val < 0.05,])
DE.serosal <- rownames(tt.serosal[tt.serosal$adj.P.Val < 0.05,])
```

# DE on "Lymphoid cells"
```{r}
# Define the cell type you want to extract
cluster.type <- "Lymphoid cells"
# Subset data and run differential expression analysis on lymphoid cells
das.control.lymphoid <- das.control_dat[, das.control.clusters == cluster.type]
das.dasatinib.lymphoid <- das.dasatinib_dat[, das.dasatinib.clusters == cluster.type]
das.combined.lymphoid <- cbind(das.control.lymphoid, das.dasatinib.lymphoid)
das.groups.tmp <- c(das.control.treat[das.control.clusters == cluster.type], das.dasatinib.treat[das.dasatinib.clusters == cluster.type])

# Define grouping variable more clearly for limma
group <- ifelse(das.groups.tmp == "Das - control", "control", "dasatinib")
# Create design matrix: looks good
design <- model.matrix(~group);
colnames(design) <- c("Mean", "ControlVsDasatinib")
design %>% head()
# Fit linear model
fit <- lmFit(as.matrix(das.combined.lymphoid), design)
fit <- eBayes(fit)
tt.lymphoid <- topTable(fit, coef = "ControlVsDasatinib", adjust = "BH", n = nrow(das.combined.lymphoid))
# Sum number of significantly DE genes in tt
sum(tt.lymphoid$adj.P.Val<0.05)
# Report names of significantly DE genes
rownames(tt.lymphoid[tt.lymphoid$adj.P.Val < 0.05,])
DE.lymphoid <- rownames(tt.lymphoid[tt.lymphoid$adj.P.Val < 0.05,])
```

# DE on "Fibroblasts"
```{r}
# Define the cell type you want to extract
cluster.type <- "Undetermined (fibroblasts)"
# Subset data and run differential expression analysis on fibroblast cells
das.control.fibroblast <- das.control_dat[, das.control.clusters == cluster.type]
das.dasatinib.fibroblast <- das.dasatinib_dat[, das.dasatinib.clusters == cluster.type]
das.combined.fibroblast <- cbind(das.control.fibroblast, das.dasatinib.fibroblast)
das.groups.tmp <- c(das.control.treat[das.control.clusters == cluster.type], das.dasatinib.treat[das.dasatinib.clusters == cluster.type])

# Define grouping variable more clearly for limma
group <- ifelse(das.groups.tmp == "Das - control", "control", "dasatinib")
# Create design matrix: looks good
design <- model.matrix(~group);
colnames(design) <- c("Mean", "ControlVsDasatinib")
design %>% head()
# Fit linear model
fit <- lmFit(as.matrix(das.combined.fibroblast), design)
fit <- eBayes(fit)
tt.fibroblast <- topTable(fit, coef = "ControlVsDasatinib", adjust = "BH", n = nrow(das.combined.fibroblast))
# Sum number of significantly DE genes in tt
sum(tt.fibroblast$adj.P.Val<0.05)
# Report names of significantly DE genes
rownames(tt.fibroblast[tt.fibroblast$adj.P.Val < 0.05,])
DE.fibroblast <- rownames(tt.fibroblast[tt.fibroblast$adj.P.Val < 0.05,])
```

# DE on "Stromal cells"
```{r}
# Define the cell type you want to extract
cluster.type <- "Stromal cells"
# Subset data and run differential expression analysis on stromal cells
das.control.stromal <- das.control_dat[, das.control.clusters == cluster.type]
das.dasatinib.stromal <- das.dasatinib_dat[, das.dasatinib.clusters == cluster.type]
das.combined.stromal <- cbind(das.control.stromal, das.dasatinib.stromal)
das.groups.tmp <- c(das.control.treat[das.control.clusters == cluster.type], das.dasatinib.treat[das.dasatinib.clusters == cluster.type])

# Define grouping variable more clearly for limma
group <- ifelse(das.groups.tmp == "Das - control", "control", "dasatinib")
# Create design matrix: looks good
design <- model.matrix(~group);
colnames(design) <- c("Mean", "ControlVsDasatinib")
design %>% head()
# Fit linear model
fit <- lmFit(as.matrix(das.combined.stromal), design)
fit <- eBayes(fit)
tt.stromal <- topTable(fit, coef = "ControlVsDasatinib", adjust = "BH", n = nrow(das.combined.stromal))
# Sum number of significantly DE genes in tt
sum(tt.stromal$adj.P.Val<0.05)
# Report names of significantly DE genes
rownames(tt.stromal[tt.stromal$adj.P.Val < 0.05,])
DE.stromal <- rownames(tt.stromal[tt.stromal$adj.P.Val < 0.05,])
```

# Create and save a table of top DE genes in each object
```{r}
# Create a list
DE.Cdh1.neg <- list(DE.macrophage, DE.lymphatic, DE.serosal, DE.lymphoid, DE.fibroblast, DE.stromal)
names(DE.Cdh1.neg) <- c("DE.macrophage", "DE.lymphatic", "DE.serosal", "DE.lymphoid", "DE.fibroblast", "DE.stromal")
capture.output(DE.Cdh1.neg, file = "2_output_strong_qc/DE.Cdh1.neg.csv")
```

# Explore differential expression between Cdh1 negative and Cdh1 positive cells within epithelial clusters
```{r}
# Define the cell types you want to extract (should be 312 controls)
ctl.fov <- which(das.control.clusters == "Foveolar cells")
ctl.pyl <- which(das.control.clusters == "Pyloric gland/neck cells")
ctl.par <- which(das.control.clusters == "Parietal cells")
# Can't use ent cause all DE genes just become Ent markers
# ctl.ent <- which(das.control.clusters == "Enterocytes")
ctl.epithelial <- c(ctl.fov, ctl.pyl, ctl.par) %>% sort(.)

# Should be 257 but there's 258? Fix this at some point but for now it seems ok.
das.fov <- which(das.dasatinib.clusters == "Foveolar cells")
das.pyl <- which(das.dasatinib.clusters == "Pyloric gland/neck cells")
das.par <- which(das.dasatinib.clusters == "Parietal cells")
# Can't use ent cause all DE genes just become Ent markers
# das.ent <- which(das.dasatinib.clusters == "Enterocytes")
das.epithelial <- c(das.fov, das.pyl, das.par) %>% sort(.)

# Subset data and run differential expression analysis on epithelial cells
# Subset to epithelial
das.control.epithelial <- das.control_dat[, ctl.epithelial]
# Subset to Cdh1- only
das.control.epithelial <- das.control.epithelial[, das.control.epithelial["Cdh1",] == 0]
# Subset to epithelial
das.dasatinib.epithelial <- das.dasatinib_dat[, das.epithelial]
# Subset to Cdh1- only
das.dasatinib.epithelial <- das.dasatinib.epithelial[, das.dasatinib.epithelial["Cdh1",] == 0]

das.combined.epithelial <- cbind(das.control.epithelial, das.dasatinib.epithelial)

das.groups.tmp <- rep(c("Das - control", "Das - drug"), c(ncol(das.control.epithelial), ncol(das.dasatinib.epithelial)))

# Define grouping variable more clearly for limma
group <- ifelse(das.groups.tmp == "Das - control", "control", "dasatinib")
# Create design matrix: looks good
design <- model.matrix(~group);
colnames(design) <- c("Mean", "ControlVsDasatinib")
design %>% head()
# Fit linear model
fit <- lmFit(as.matrix(das.combined.epithelial), design)
fit <- eBayes(fit)
tt.epithelial <- topTable(fit, coef = "ControlVsDasatinib", adjust = "BH", n = nrow(das.combined.epithelial))
# Sum number of significantly DE genes in tt
sum(tt.epithelial$adj.P.Val<0.05)
# Report names of significantly DE genes
rownames(tt.epithelial[tt.epithelial$adj.P.Val < 0.05,])
DE.epithelial <- rownames(tt.epithelial[tt.epithelial$adj.P.Val < 0.05,])
```

# Explore all epithelial cells, not just cdh1-
```{r}
# Define the cell types you want to extract (should be 312 controls)
ctl.fov <- which(das.control.clusters == "Foveolar cells")
ctl.pyl <- which(das.control.clusters == "Pyloric gland/neck cells")
ctl.par <- which(das.control.clusters == "Parietal cells")
# Can't use ent cause all DE genes just become Ent markers
# ctl.ent <- which(das.control.clusters == "Enterocytes")
ctl.epithelial.all <- c(ctl.fov, ctl.pyl, ctl.par) %>% sort(.)

# Should be 257 but there's 258? Fix this at some point but for now it seems ok.
das.fov <- which(das.dasatinib.clusters == "Foveolar cells")
das.pyl <- which(das.dasatinib.clusters == "Pyloric gland/neck cells")
das.par <- which(das.dasatinib.clusters == "Parietal cells")
# Can't use ent cause all DE genes just become Ent markers
# das.ent <- which(das.dasatinib.clusters == "Enterocytes")
das.epithelial.all <- c(das.fov, das.pyl, das.par) %>% sort(.)

# Subset data and run differential expression analysis on epithelial.all cells
# Subset to epithelial.all
das.control.epithelial.all <- das.control_dat[, ctl.epithelial.all]
# Subset to epithelial
das.dasatinib.epithelial.all <- das.dasatinib_dat[, das.epithelial.all]

das.combined.epithelial.all <- cbind(das.control.epithelial.all, das.dasatinib.epithelial.all)

das.groups.tmp <- rep(c("Das - control", "Das - drug"), c(ncol(das.control.epithelial.all), ncol(das.dasatinib.epithelial.all)))

# Define grouping variable more clearly for limma
group <- ifelse(das.groups.tmp == "Das - control", "control", "dasatinib")
# Create design matrix: looks good
design <- model.matrix(~group);
colnames(design) <- c("Mean", "ControlVsDasatinib")
design %>% head()
# Fit linear model
fit <- lmFit(as.matrix(das.combined.epithelial.all), design)
fit <- eBayes(fit)
tt.epithelial.all <- topTable(fit, coef = "ControlVsDasatinib", adjust = "BH", n = nrow(das.combined.epithelial.all))
# Sum number of significantly DE genes in tt
sum(tt.epithelial.all$adj.P.Val<0.05)
# Report names of significantly DE genes
rownames(tt.epithelial.all[tt.epithelial.all$adj.P.Val < 0.05,])
DE.epithelial.all <- rownames(tt.epithelial.all[tt.epithelial.all$adj.P.Val < 0.05,])

# Pull out which of these are significant
sig.tt.epithelial.all <- tt.epithelial.all[tt.epithelial.all$adj.P.Val < 0.05,]
# Sum how many are up in response to dasatinib
nrow(sig.tt.epithelial.all[sig.tt.epithelial.all$logFC > 0,])
# Sum how many are down in response to dasatinib
rownames(sig.tt.epithelial.all)[sig.tt.epithelial.all$logFC < 0] %>% cat()

write.csv(rownames(sig.tt.epithelial.all)[sig.tt.epithelial.all$logFC < 0], file = "2_output_strong_qc/epithelial_down.csv")
write.csv(rownames(sig.tt.epithelial.all), file = "2_output_strong_qc/epithelial_up_down.csv")


write.csv(rownames(sig.tt.epithelial.all)[sig.tt.epithelial.all$logFC > 0], file = "2_output_strong_qc/epithelial_up.csv")
```

# Run DE analysis on Cdh1 positive clusters as a whole (i.e. epithelial cells)
# DE on "Foveolar cells"
```{r}
# Define the cell type you want to extract
cluster.type <- "Foveolar cells"
# Subset data and run differential expression analysis on foveolar cells
das.control.foveolar <- das.control_dat[, das.control.clusters == cluster.type]
das.dasatinib.foveolar <- das.dasatinib_dat[, das.dasatinib.clusters == cluster.type]
das.combined.foveolar <- cbind(das.control.foveolar, das.dasatinib.foveolar)
das.groups.tmp <- c(das.control.treat[das.control.clusters == cluster.type], das.dasatinib.treat[das.dasatinib.clusters == cluster.type])

# Define grouping variable more clearly for limma
group <- ifelse(das.groups.tmp == "Das - control", "control", "dasatinib")
# Create design matrix: looks good
design <- model.matrix(~group);
colnames(design) <- c("Mean", "ControlVsDasatinib")
design %>% head()
# Fit linear model
fit <- lmFit(as.matrix(das.combined.foveolar), design)
fit <- eBayes(fit)
tt.foveolar <- topTable(fit, coef = "ControlVsDasatinib", adjust = "BH", n = nrow(das.combined.foveolar))
# Sum number of significantly DE genes in tt
sum(tt.foveolar$adj.P.Val<0.05)
# Report names of significantly DE genes
rownames(tt.foveolar[tt.foveolar$adj.P.Val < 0.05,])
DE.foveolar <- rownames(tt.foveolar[tt.foveolar$adj.P.Val < 0.05,])
```

# DE on "Pyloric gland/neck cells"
```{r}
# Define the cell type you want to extract
cluster.type <- "Pyloric gland/neck cells"
# Subset data and run differential expression analysis on pyloric gland/neck cells
das.control.pyloric <- das.control_dat[, das.control.clusters == cluster.type]
das.dasatinib.pyloric <- das.dasatinib_dat[, das.dasatinib.clusters == cluster.type]
das.combined.pyloric <- cbind(das.control.pyloric, das.dasatinib.pyloric)
das.groups.tmp <- c(das.control.treat[das.control.clusters == cluster.type], das.dasatinib.treat[das.dasatinib.clusters == cluster.type])

# Define grouping variable more clearly for limma
group <- ifelse(das.groups.tmp == "Das - control", "control", "dasatinib")
# Create design matrix: looks good
design <- model.matrix(~group);
colnames(design) <- c("Mean", "ControlVsDasatinib")
design %>% head()
# Fit linear model
fit <- lmFit(as.matrix(das.combined.pyloric), design)
fit <- eBayes(fit)
tt.pyloric <- topTable(fit, coef = "ControlVsDasatinib", adjust = "BH", n = nrow(das.combined.pyloric))
# Sum number of significantly DE genes in tt
sum(tt.pyloric$adj.P.Val<0.05)
# Report names of significantly DE genes
rownames(tt.pyloric[tt.pyloric$adj.P.Val < 0.05,])
DE.pyloric <- rownames(tt.pyloric[tt.pyloric$adj.P.Val < 0.05,])
```

# DE on "Pancreatic cells"
```{r}
# Define the cell type you want to extract
cluster.type <- "Pancreatic cells"
# Subset data and run differential expression analysis on pancreatic cells
das.control.pancreatic <- das.control_dat[, das.control.clusters == cluster.type]
das.dasatinib.pancreatic <- das.dasatinib_dat[, das.dasatinib.clusters == cluster.type]
das.combined.pancreatic <- cbind(das.control.pancreatic, das.dasatinib.pancreatic)
das.groups.tmp <- c(das.control.treat[das.control.clusters == cluster.type], das.dasatinib.treat[das.dasatinib.clusters == cluster.type])

# Define grouping variable more clearly for limma
group <- ifelse(das.groups.tmp == "Das - control", "control", "dasatinib")
# Create design matrix: looks good
design <- model.matrix(~group);
colnames(design) <- c("Mean", "ControlVsDasatinib")
design %>% head()
# Fit linear model
fit <- lmFit(as.matrix(das.combined.pancreatic), design)
fit <- eBayes(fit)
tt.pancreatic <- topTable(fit, coef = "ControlVsDasatinib", adjust = "BH", n = nrow(das.combined.pancreatic))
# Sum number of significantly DE genes in tt
sum(tt.pancreatic$adj.P.Val<0.05)
# Report names of significantly DE genes
rownames(tt.pancreatic[tt.pancreatic$adj.P.Val < 0.05,])
DE.pancreatic <- rownames(tt.pancreatic[tt.pancreatic$adj.P.Val < 0.05,])
```

# DE on "Parietal cells"
```{r}
# Define the cell type you want to extract
cluster.type <- "Parietal cells"
# Subset data and run differential expression analysis on parietal cells
das.control.parietal <- das.control_dat[, das.control.clusters == cluster.type]
das.dasatinib.parietal <- das.dasatinib_dat[, das.dasatinib.clusters == cluster.type]
das.combined.parietal <- cbind(das.control.parietal, das.dasatinib.parietal)
das.groups.tmp <- c(das.control.treat[das.control.clusters == cluster.type], das.dasatinib.treat[das.dasatinib.clusters == cluster.type])

# Define grouping variable more clearly for limma
group <- ifelse(das.groups.tmp == "Das - control", "control", "dasatinib")
# Create design matrix: looks good
design <- model.matrix(~group);
colnames(design) <- c("Mean", "ControlVsDasatinib")
design %>% head()
# Fit linear model
fit <- lmFit(as.matrix(das.combined.parietal), design)
fit <- eBayes(fit)
tt.parietal <- topTable(fit, coef = "ControlVsDasatinib", adjust = "BH", n = nrow(das.combined.parietal))
# Sum number of significantly DE genes in tt
sum(tt.parietal$adj.P.Val<0.05)
# Report names of significantly DE genes
rownames(tt.parietal[tt.parietal$adj.P.Val < 0.05,])
DE.parietal <- rownames(tt.parietal[tt.parietal$adj.P.Val < 0.05,])
```

# Write out all toptables so that you have them saved as files for posterity
```{r}
# tt.epithelial.all
write.csv(tt.epithelial.all, file = "2_output_strong_qc/tt_epithelial.csv")

# tt.fibroblast
write.csv(tt.fibroblast, file = "2_output_strong_qc/tt_fibroblast.csv")

# tt.foveolar
write.csv(tt.foveolar, file = "2_output_strong_qc/tt_foveolar.csv")

# tt.lymphatic
write.csv(tt.lymphatic, file = "2_output_strong_qc/tt_lymphatic.csv")

# tt.lymphoid
write.csv(tt.lymphoid, file = "2_output_strong_qc/tt_lymphoid.csv")

# tt.macrophage
write.csv(tt.macrophage, file = "2_output_strong_qc/tt_macrophage.csv")

# tt.pancreatic
write.csv(tt.pancreatic, file = "2_output_strong_qc/tt_pancreatic.csv")

# tt.parietal
write.csv(tt.parietal, file = "2_output_strong_qc/tt_parietal.csv")

# tt.pyloric
write.csv(tt.pyloric, file = "2_output_strong_qc/tt_pyloric.csv")

# tt.serosal
write.csv(tt.serosal, file = "2_output_strong_qc/tt_serosal.csv")

# tt.stromal
write.csv(tt.stromal, file = "2_output_strong_qc/tt_stromal.csv")

```